UserType에 따라 작성 가능한 게시물을 제한하기 위한 로직 작성 및 그에 따른 테스트 코드 수정

hamryt opened the pull request:
이슈번호 : #57

hamryt committed:
UserType에 따라 작성 가능한 게시물을 제한하기 위한 로직 작성 및 그에 따른 테스트 코드 수정

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/dto/UserType.java:
```diff
@@ -10,4 +12,10 @@
         this.name = name;
     }
+    
+    public static void checkUserType(UserType userType, UserType authType){
```
1. `void checkUserType(UserType userType, UserType authType)`라는 모습을 봤을 때 어떤 역할의 메소드인지 파악이 가능할까요?
2. 이 메소드의 목적을 보았을 때 이 클래스에 있는게 맞을까요?
3. 이런 역할의 경우 `check~~`라는 이름보다는 `validate~~`로 지어주는게 좋습니다

hamryt committed:
유저 타입 검증 로직 메서드를 단순 비교만 하기에 static으로 하고 서비스 단에서 클래스들의 순환참조를 방지하기 위해 별도의 클래스로 분리

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/util/UserTypeUtil.java:
```diff
@@ -0,0 +1,17 @@
+    private UserTypeUtil() {
+    }
+    
+    public static void validateUserType(UserType userType, UserType authType) {
```
`validateUserType(UserType userType, UserType authType)`라는 시그니처 또한 어떤 것을 검사하는건지 명확하지 않습니다.

이 메소드의 역할은 두 인자가 같은지 검사한다라는 사실은 동호님만 아는 사실입니다. 이 사실을 모르는 사람에게 어떻게 효율적으로 시그니처만으로 전달할 수 있을지 고민해주세요

hamryt committed:
유저타입 검사 메서드 이름 변경, 예외 클래스도 사용 목적을 알기 쉽게 구조 변경

hamryt committed:
merge develop

hamryt committed:
InsertGymFailedException 인식 못하는 문제 해결

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/util/UserTypeUtil.java:
```diff
@@ -1,16 +1,16 @@
     }
     
-    public static void validateUserType(UserType userType, UserType authType) {
+    public static void validateUserTypePermissions(UserType userType, UserType authType) {
```
유저 권한을 검사한다라는건 동호님이 이 메소드를 사용하려는 목적이지 이 메소드의 동작을 설명해주는 이름은 아닌 것 같습니다.
이 메소드의 내부 동작만 보았을 때 이 메소드는 어떤 동작을 수행하는 메소드인가요? 메소드 시그니처는 그것을 명확하게 보여주고 있나요?

hamryt replied:
제가 지정한 이름은 제가 어떤 의도를 가지고 메서드를 만들었는가 였습니다. 그래서 메서드에는 무엇을 하는지 보다는 이것이 정의된 목적을 썼습니다. 멘토님의 말씀을 보면 메서드는 동작 목적보다는 방식이 더 중요한 네이밍 요소인 것 같습니다. 그리고 그런 이유는 아마 다른 사람들과 개발을 할 때 해당 메서드가 어떤 동작을 하는지 이름에서 잘 명시하고 있어야 코드의 가독성이 좋아지기 때문일거라 생각합니다. 그래서 제가 만든 메서드도 이름은 validEqualUserType으로 바꾸고 메서드 위치를 enum클래스 안으로 옮긴 다음 해당 객체가 가지고 있는 값을 가지고 비교하게 해서 추가로 메서드를 담을 클래스를 만들지 않게하였습니다. 그리고 두개의 비교 유저타입 파라미터를 전달하지 않고 게시판의 유저타입만 전달하고 enum 객체가 가지고 있는 값과 비교하게 바꾸었습니다.

hamryt committed:
userType 비교 메서드 enum 클래스로 위치 변경
이유 : 요청 객체 안의 enum 객체의 값으로 본인 클래스 안에서 비교하면 추가 클래스를 만들지 않아도 되고 본 객체가 가지고 있는 값으로 비교만 하면 되기 때문에 전달하는 인자도 1로 준다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/dto/UserType.java:
```diff
@@ -1,14 +1,25 @@
+    public void validEqualUserType(String authType){
+        if (!name.equals(authType)){
+            log.error("UserType of request does not match with UserType of board! userTypeRequest: {}, userTypeBoard: [}", name, authType);
+            throw new PermissionException("This UserType does not appropriate for the board ");
```
이름이 이제 적절하네요~ 👍

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/dto/UserType.java:
```diff
@@ -1,14 +1,25 @@
     
+    public void validEqualUserType(String authType){
+        if (!name.equals(authType)){
+            log.error("UserType of request does not match with UserType of board! userTypeRequest: {}, userTypeBoard: [}", name, authType);
```
로그는 굳이 남길 필요는 없어보입니다. 아래 예외 메세지를 조금 더 자세하게 해주시는게 나을 것 같아요

hamryt committed:
유저 타입 불일치 예외 수정

hamryt committed:
예외 메세지 수정한 것에 맞춰 테스트 코드 수정

hamryt committed:
pull develop
